### PR TITLE
Basic support for website creation

### PR DESCRIPTION
**Pull request by fishnix**
Supports:
* creating the bucket
* setting the website configuration
* configuring the bucket policy for public access

Does not yet support:
* creating a dns record for the site
* assigning the SSL certificate
* setting up the cloudfront distribution

Additionally:
* Update readme with all endpoints
* Change Tags to be the native AWS format instead of one that makes sense 😆 
* change the service ports for k8s to be istio compliant
* introduce a better way to rollback without panicing which will be applied to bucket creation as well

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -341,6 +362,117 @@ DELETE `/v1/s3/{account}/buckets/{bucket}/users/{user}
 | **429 Too Many Requests**     | service or rate limit exceeded           |  
 | **500 Internal Server Error** | a server error occurred                  |
 
+### Create a website
+
+POST `/v1/s3/{account}/websites
```
Comment (tenyo) [inline README.md]: 👀 closing `

File: README.md [hunk 2]
```diff
@@ -341,6 +362,117 @@ DELETE `/v1/s3/{account}/buckets/{bucket}/users/{user}
...
+        "Arn": "arn:aws:iam::846761448161:policy/foobar.bulldogs.cloud-BktAdmPlc",
+        "AttachmentCount": 0,
+        "CreateDate": "2019-03-01T15:33:52Z",
+        "DefaultVersionId": "v1",
+        "Description": null,
+        "IsAttachable": true,
+        "Path": "/",
+        "PermissionsBoundaryUsageCount": 0,
+        "PolicyId": "ABCDEFGHI12345678",
+        "PolicyName": "foobar.bulldogs.cloud-BktAdmPlc",
```
Comment (tenyo) [inline README.md]: What's the length limit for `PolicyName` and `GroupName`?

File: README.md [hunk 3]
```diff
@@ -341,6 +362,117 @@ DELETE `/v1/s3/{account}/buckets/{bucket}/users/{user}
...
+
+### Delete a website
+
+DELETE `/v1/s3/{account}/websites/{website}`
+
+*See [Delete a bucket](#delete-a-bucket)*
+
+### Create a website user
+
+POST `/v1/s3/{account}/websites/{website}/users
```
Comment (tenyo) [inline README.md]: 👀

File: README.md [hunk 4]
```diff
@@ -341,6 +362,117 @@ DELETE `/v1/s3/{account}/buckets/{bucket}/users/{user}
...
+
+*See [Create a bucket user](#create-a-bucket-user)*
+
+### List users for a website
+
+GET `/v1/s3/{account}/websites/{website}/users/{user}
```
Comment (tenyo) [inline README.md]: 👀

File: s3/buckets.go [hunk 5]
```diff
@@ -172,3 +163,53 @@ func (s *S3) TagBucket(ctx context.Context, bucket string, tags map[string]strin
...
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			default:
+				return apierror.New(apierror.ErrBadRequest, aerr.Message(), err)
+			}
+		}
+
+		return apierror.New(apierror.ErrInternalError, "unknown error occurred", err)
+	}
+	return err
```
Comment (tenyo) [inline s3/buckets.go]: This will always return `nil`, not sure if it's clearer to just `return nil`

File: s3/buckets.go [hunk 6]
```diff
@@ -172,3 +163,53 @@ func (s *S3) TagBucket(ctx context.Context, bucket string, tags map[string]strin
...
```
Comment (tenyo) [inline s3/buckets.go]: same

File: s3api/handlers_website.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+		GroupName: aws.String(groupName),
+	})
+
+	if err != nil {
+		msg := fmt.Sprintf("failed to create group: %s", err.Error())
+		handleError(w, errors.Wrap(err, msg))
+		return
+	}
+
+	// append policy delete to rollback tasks
```
Comment (tenyo) [inline s3api/handlers_website.go]: 👀 group delete

File: s3api/handlers_website.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+			return nil
+		}()
+	}
+	rollBackTasks = append(rollBackTasks, rbfunc)
+
+	if _, err = iamService.AttachGroupPolicy(r.Context(), &iam.AttachGroupPolicyInput{
+		GroupName: aws.String(groupName),
+		PolicyArn: policyOutput.Policy.Arn,
+	}); err != nil {
+		msg := fmt.Sprintf("failed to create group: %s", err.Error())
```
Comment (tenyo) [inline s3api/handlers_website.go]: 👀 attach group policy

File: README.md [hunk 9]
```diff
@@ -341,6 +362,117 @@ DELETE `/v1/s3/{account}/buckets/{bucket}/users/{user}
...
+
+*See [Reset access keys for a bucket user](#reset-access-keys-for-a-bucket-user)*
+
+### Delete a website user
+
+DELETE `/v1/s3/{account}/websites/{website}/users/{user}`
+
+*See [Delete a bucket user](#delete-a-bucket-user)*
+
+
```
Comment (tenyo) [inline README.md]: excellent readme 😸 👍

File: s3api/handlers_website.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+		handleError(w, apierror.New(apierror.ErrBadRequest, msg, err))
+		return
+	}
+
+	// setup err var, rollback function list and defer execution
+	var rollBackTasks []func() error
+	defer func() {
+		if err != nil {
+			log.Errorf("recovering from error: %s, executing %d rollback tasks", err, len(rollBackTasks))
+			rollBack(&rollBackTasks)
```
Comment (tenyo) [inline s3api/handlers_website.go]: 💯 nice!

File: s3api/handlers_website.go [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write(j)
+}
+
+// rollBack executes functions from a stack of rollback functions
+func rollBack(t *[]func() error) {
```
Comment (tenyo) [inline s3api/handlers_website.go]: This will be shared, right? Can probably move to another file (`server.go`)

Comment (fishnix) [reply s3api/handlers_website.go]: yeah definately, when i do the reimplementation of the rollback for buckets, ill move it

File: README.md (hunk 2 above)
Comment (fishnix) [reply README.md]: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-limits.html#reference_iam-limits-entity-length

File: s3/buckets.go (hunk 6 above)
Comment (fishnix) [reply s3/buckets.go]: 👍

File: README.md (hunk 9 above)
Comment (fishnix) [reply README.md]: 🎉

File: README.md (hunk 2 above)
Comment (tenyo) [reply README.md]: ok, 128 should be plenty

**Comment by tenyo** [comment]
Looks good 👍

**Pull request merged**